File: app/src/main/java/com/arman/queuetube/activities/MainActivity.java
package com.arman.queuetube.activities;

import android.content.Intent;
import android.content.IntentFilter;
import android.content.res.Configuration;
import android.net.ConnectivityManager;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;

import com.arman.queuetube.R;
import com.arman.queuetube.config.Constants;
import com.arman.queuetube.fragments.MainFragment;
import com.arman.queuetube.fragments.PlayerFragment;
import com.arman.queuetube.fragments.PlaylistFragment;
import com.arman.queuetube.fragments.PlaylistsFragment;
import com.arman.queuetube.fragments.StreamFragment;
import com.arman.queuetube.listeners.DrawerItemListener;
import com.arman.queuetube.model.VideoData;
import com.arman.queuetube.modules.playlists.JSONPlaylistHelper;
import com.arman.queuetube.util.notifications.receivers.WifiReceiver;
import com.arman.queuetube.util.services.KillNotificationService;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.MobileAds;
import com.google.android.material.navigation.NavigationView;

import java.util.Collection;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.FragmentTransaction;
import androidx.viewpager.widget.ViewPager;

public class MainActivity extends AppCompatActivity
        implements PlaylistFragment.OnPlayItemsListener {

    public static final String TAG = "MainActivity";

    public static final int MAIN_FRAGMENT = 0;
    public static final int STREAM_FRAGMENT = 1;
    public static final int FAVORITES_FRAGMENT = 2;
    public static final int HISTORY_FRAGMENT = 3;
    public static final int PLAYLISTS_FRAGMENT = 4;

    private DrawerLayout drawerLayout;
    private ActionBarDrawerToggle drawerToggle;
    private LinearLayout extendedToolbar;

    private MainFragment mainFragment;
    private PlaylistFragment favoritesFragment;
    private PlaylistFragment historyFragment;
    private StreamFragment streamFragment;
    private PlaylistsFragment playlistsFragment;

    private WifiReceiver wifiReceiver;

    private int currentFragment;

    private void setupWifiReceiver() {
        this.wifiReceiver = new WifiReceiver();
        IntentFilter filter = new IntentFilter();
        filter.addAction(ConnectivityManager.CONNECTIVITY_ACTION);
        registerReceiver(this.wifiReceiver, filter);
    }

    private void setupDrawer() {
        this.drawerLayout = (DrawerLayout) findViewById(R.id.drawer_layout);

        NavigationView navigationView = this.drawerLayout.findViewById(R.id.drawer_navigation_view);
        navigationView.setCheckedItem(R.id.menu_item_home);
        DrawerItemListener drawerListener = new DrawerItemListener(this, this.drawerLayout);
        navigationView.setNavigationItemSelectedListener(drawerListener);
        this.drawerLayout.addDrawerListener(drawerListener);
    }

    private void setupActionBar() {
        Toolbar toolbar = (Toolbar) this.findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setHomeAsUpIndicator(R.drawable.ic_menu_black_24dp);

        this.setupDrawerToggle(toolbar);
    }

    private void setupDrawerToggle(Toolbar toolbar) {
        this.drawerToggle = new ActionBarDrawerToggle(this, this.drawerLayout, toolbar, R.string.drawer_open, R.string.drawer_close);
        this.drawerLayout.addDrawerListener(this.drawerToggle);
    }

    private void setupAdView() {
        MobileAds.initialize(this, Constants.Key.TEST_AD_KEY);
        AdView adView = (AdView) this.findViewById(R.id.ad_view);
        AdRequest adRequest = new AdRequest.Builder().build();
        adView.loadAd(adRequest);
    }

    private void setupPlaylistFragment(int index) {
        Bundle bundle = new Bundle();
        switch (index) {
            case FAVORITES_FRAGMENT:
                bundle.putString("playlistName", JSONPlaylistHelper.FAVORITES);
                this.favoritesFragment = new PlaylistFragment();
                this.favoritesFragment.setArguments(bundle);
                this.favoritesFragment.setOnPlayItemsListener(this);
                break;
            case HISTORY_FRAGMENT:
                bundle.putString("playlistName", JSONPlaylistHelper.HISTORY);
                bundle.putBoolean("isClickable", false);
                this.historyFragment = new PlaylistFragment();
                this.historyFragment.setArguments(bundle);
                this.historyFragment.setOnPlayItemsListener(this);
                break;
            default:
                break;
        }
    }

    private void setupFragments() {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();

        this.setupPlaylistFragment(FAVORITES_FRAGMENT);
        transaction.replace(R.id.favorites_fragment_frame, this.favoritesFragment).hide(this.favoritesFragment);

        this.setupPlaylistFragment(HISTORY_FRAGMENT);
        transaction.replace(R.id.history_fragment_frame, this.historyFragment).hide(this.historyFragment);

//        this.playlistsFragment = new PlaylistsFragment();
//        transaction.replace(R.id.playlists_fragment_frame, this.playlistsFragment).hide(this.playlistsFragment);

        this.streamFragment = new StreamFragment();
        transaction.replace(R.id.stream_fragment_frame, this.streamFragment).hide(this.streamFragment);

        transaction.commit();

        this.switchToMainFragment();
    }

    public void switchToMainFragment() {
        if (this.currentFragment != MAIN_FRAGMENT) {
            FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
            if (this.mainFragment != null) {
                transaction.show(this.mainFragment);
                this.refreshVideoFavorited();
            } else {
                this.mainFragment = new MainFragment();
                transaction.replace(R.id.main_fragment_frame, this.mainFragment);
            }
            if (this.favoritesFragment != null) {
                transaction.hide(this.favoritesFragment);
            }
            if (this.historyFragment != null) {
                transaction.hide(this.historyFragment);
            }
//            if (this.playlistsFragment != null) {
//                transaction.hide(this.playlistsFragment);
//            }
            if (this.streamFragment != null) {
                transaction.hide(this.streamFragment);
            }
            if (this.extendedToolbar != null) {
                this.extendedToolbar.setVisibility(View.VISIBLE);
            }
            transaction.commitNow();
            this.currentFragment = MAIN_FRAGMENT;
        }
    }

    public void switchToFavoritesFragment() {
        if (this.currentFragment != FAVORITES_FRAGMENT) {
            FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
            if (this.favoritesFragment != null) {
                transaction.show(this.favoritesFragment);
            } else {
                this.setupPlaylistFragment(FAVORITES_FRAGMENT);
                transaction.replace(R.id.favorites_fragment_frame, this.favoritesFragment);
            }
            if (this.mainFragment != null) {
                transaction.hide(this.mainFragment);
            }
            if (this.historyFragment != null) {
                transaction.hide(this.historyFragment);
            }
            if (this.streamFragment != null) {
                transaction.hide(this.streamFragment);
            }
            if (this.extendedToolbar != null) {
                this.extendedToolbar.setVisibility(View.GONE);
            }
            transaction.commitNow();
            this.currentFragment = FAVORITES_FRAGMENT;
        }
    }

    public void switchToHistoryFragment() {
        if (this.currentFragment != HISTORY_FRAGMENT) {
            FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
            if (this.historyFragment != null) {
                transaction.show(this.historyFragment);
            } else {
                this.setupPlaylistFragment(HISTORY_FRAGMENT);
                transaction.replace(R.id.history_fragment_frame, this.historyFragment);
            }
            if (this.mainFragment != null) {
                transaction.hide(this.mainFragment);
            }
            if (this.favoritesFragment != null) {
                transaction.hide(this.favoritesFragment);
            }
            if (this.streamFragment != null) {
                transaction.hide(this.streamFragment);
            }
            if (this.extendedToolbar != null) {
                this.extendedToolbar.setVisibility(View.GONE);
            }
            transaction.commitNow();
            this.currentFragment = HISTORY_FRAGMENT;
        }
    }

    public void switchToPlaylistsFragment() {
        if (this.currentFragment != PLAYLISTS_FRAGMENT) {
            FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
            if (this.playlistsFragment != null) {
                transaction.show(this.playlistsFragment);
            } else {
                this.playlistsFragment = new PlaylistsFragment();
                transaction.replace(R.id.playlists_fragment_frame, this.playlistsFragment);
            }
            if (this.mainFragment != null) {
                transaction.hide(this.mainFragment);
            }
            if (this.favoritesFragment != null) {
                transaction.hide(this.favoritesFragment);
            }
            if (this.historyFragment != null) {
                transaction.hide(this.historyFragment);
            }
            if (this.streamFragment != null) {
                transaction.hide(this.streamFragment);
            }
            if (this.extendedToolbar != null) {
                this.extendedToolbar.setVisibility(View.GONE);
            }
            transaction.commitNow();
            this.currentFragment = PLAYLISTS_FRAGMENT;
        }
    }

    public void switchToStreamFragment() {
        if (this.currentFragment != STREAM_FRAGMENT) {
            FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
            if (this.streamFragment != null) {
                transaction.show(this.streamFragment);
            } else {
                this.streamFragment = new StreamFragment();
                transaction.replace(R.id.stream_fragment_frame, this.streamFragment);
            }
            if (this.mainFragment != null) {
                transaction.hide(this.mainFragment);
            }
            if (this.favoritesFragment != null) {
                transaction.hide(this.favoritesFragment);
            }
            if (this.historyFragment != null) {
                transaction.hide(this.historyFragment);
            }
//            if (this.playlistsFragment != null) {
//                transaction.hide(this.playlistsFragment);
//            }
            if (this.extendedToolbar != null) {
                this.extendedToolbar.setVisibility(View.GONE);
            }
            transaction.commitNow();
            this.currentFragment = STREAM_FRAGMENT;
        }
    }

    public void refreshVideoFavorited() {
        PlayerFragment playerFragment = (PlayerFragment) this.mainFragment.getPagerAdapter().getPlayerFragment();
        VideoData currentVideo = playerFragment.getCurrentVideo();
        playerFragment.updateVideo(JSONPlaylistHelper.isFavorited(currentVideo));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setTheme(R.style.AppTheme);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        JSONPlaylistHelper.onCreate(this);

        this.currentFragment = -1;

        startService(new Intent(this, KillNotificationService.class));

        this.extendedToolbar = (LinearLayout) findViewById(R.id.main_extended_toolbar);

        setupWifiReceiver();
        setupDrawer();
        setupAdView();
        setupActionBar();
        setupFragments();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (this.drawerToggle.onOptionsItemSelected(item)) {
            return true;
        }

        switch (item.getItemId()) {
            case R.id.action_search:
                return true;
            case android.R.id.home:
                this.drawerLayout.openDrawer(GravityCompat.START);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onBackPressed() {
        if (this.drawerLayout.isDrawerOpen(GravityCompat.START)) {
            this.drawerLayout.closeDrawer(GravityCompat.START);
        } else if (this.currentFragment == MAIN_FRAGMENT) {
            ViewPager viewPager = this.mainFragment.getViewPager();
            if (viewPager.getCurrentItem() == 0) {
                super.onBackPressed();
            } else {
                viewPager.setCurrentItem(viewPager.getCurrentItem() - 1);
            }
        } else {
            NavigationView navigationView = (NavigationView) this.drawerLayout.findViewById(R.id.drawer_navigation_view);
            navigationView.getMenu().findItem(R.id.menu_item_home).setChecked(true);
            this.switchToMainFragment();
        }
    }

    @Override
    protected void onPostCreate(@Nullable Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        this.drawerToggle.syncState();
    }

    @Override
    public void onConfigurationChanged(@NonNull Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        this.drawerToggle.onConfigurationChanged(newConfig);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        unregisterReceiver(this.wifiReceiver);
    }

    public void refreshPlaylists() {
        if (this.playlistsFragment != null) {
            this.playlistsFragment.loadPlaylists();
        }
        if (this.favoritesFragment != null) {
            this.favoritesFragment.loadPlaylist();
        }
        if (this.historyFragment != null) {
            this.historyFragment.loadPlaylist();
        }
    }

    @Override
    public void onPlayAll(Collection<VideoData> videos) {
        PlayerFragment playerFragment = (PlayerFragment) this.mainFragment.getPagerAdapter().getPlayerFragment();
        if (playerFragment.setQueueTo(videos)) {
            playerFragment.forcePlayNext();
        }
    }

    @Override
    public void onPlay(VideoData video) {
        PlayerFragment playerFragment = (PlayerFragment) this.mainFragment.getPagerAdapter().getPlayerFragment();
        if (playerFragment.addToQueue(video)) {
            playerFragment.tryPlayNext();
        }
    }

}

File: app/src/main/java/com/arman/queuetube/activities/MainActivity.kt
package com.arman.queuetube.activities

import android.content.ActivityNotFoundException
import android.content.Intent
import android.content.IntentFilter
import android.net.ConnectivityManager
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.arman.queuetube.R
import com.arman.queuetube.config.Constants
import com.arman.queuetube.fragments.main.HomeFragment
import com.arman.queuetube.fragments.main.LibraryFragment
import com.arman.queuetube.fragments.main.PlayerFragment
import com.arman.queuetube.fragments.main.SearchFragment
import com.arman.queuetube.listeners.OnPlayItemsListener
import com.arman.queuetube.model.Video
import com.arman.queuetube.modules.playlists.json.GsonPlaylistHelper
import com.arman.queuetube.modules.search.YouTubeService
import com.arman.queuetube.receivers.PlayReceiver
import com.arman.queuetube.receivers.WifiReceiver
import com.arman.queuetube.services.KillNotificationService
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.bottomnavigation.BottomNavigationView
import kotlinx.android.synthetic.main.activity_main.*
import java.util.regex.Pattern


class MainActivity : AppCompatActivity(), OnPlayItemsListener, BottomNavigationView.OnNavigationItemSelectedListener {

    private var playerFragment: PlayerFragment? = null

    private var homeFragment: HomeFragment? = null
    private var searchFragment: SearchFragment? = null
    private var libraryFragment: LibraryFragment? = null

    private var wifiReceiver: WifiReceiver? = null
    private var playReceiver: PlayReceiver? = null

    private var currentFragment: Int = 0

    private var savedInstanceState: Bundle? = null

    private var bottomNavUp: Boolean = true

    fun enableScroll(): Int {
        val params = toolbar.layoutParams as AppBarLayout.LayoutParams
        val flags = params.scrollFlags
        params.scrollFlags =
                AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL or
                        AppBarLayout.LayoutParams.SCROLL_FLAG_ENTER_ALWAYS or
                        AppBarLayout.LayoutParams.SCROLL_FLAG_SNAP
        toolbar.layoutParams = params
        return flags
    }

    fun setScroll(scrollFlags: Int) {
        val params = toolbar.layoutParams as AppBarLayout.LayoutParams
        params.scrollFlags = scrollFlags
        toolbar.layoutParams = params
    }

    fun disableScroll() {
        val params = toolbar.layoutParams as AppBarLayout.LayoutParams
        params.scrollFlags = 0
        toolbar.layoutParams = params
    }

    fun hideBottomBar() {
        if (bottomNavUp) {
            with(bottom_nav_bar, {
                animate()
                        .translationY(height.toFloat())
                        .alpha(0f)
                        .withEndAction {
                            visibility = View.GONE
                            bottomNavUp = false
                        }
            })
        }
    }

    fun showBottomBar() {
        if (!bottomNavUp) {
            with(bottom_nav_bar, {
                visibility = View.VISIBLE
                animate()
                        .translationY(0f)
                        .alpha(1f)
                        .withEndAction {
                            bottomNavUp = true
                        }
            })
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        this.savedInstanceState = outState
    }

    private fun setupReceivers() {
        this.wifiReceiver = WifiReceiver()
        var filter = IntentFilter()
        filter.addAction(ConnectivityManager.CONNECTIVITY_ACTION)
        registerReceiver(this.wifiReceiver, filter)

        this.playReceiver = PlayReceiver(this)
        filter = IntentFilter()
        filter.addAction(Constants.Action.Play.PLAY_ACTION)
        filter.addAction(Constants.Action.Play.PLAY_NEXT_ACTION)
        filter.addAction(Constants.Action.Play.PLAY_NOW_ACTION)
        filter.addAction(Constants.Action.Play.PLAY_ALL_ACTION)
        filter.addAction(Constants.Action.Play.SHUFFLE_ACTION)
        registerReceiver(this.playReceiver, filter)
    }

    private fun setupPlayerFragment() {
        val transaction = supportFragmentManager.beginTransaction()
        when {
            this.playerFragment != null -> {
                transaction.show(this.playerFragment!!)
                this.refreshVideoFavorited()
            }
            else -> {
                this.playerFragment = supportFragmentManager.findFragmentByTag(PlayerFragment.TAG) as PlayerFragment?
                if (this.playerFragment != null) {
                    transaction.show(this.playerFragment!!)
                    this.refreshVideoFavorited()
                } else {
                    this.playerFragment = PlayerFragment()
                    transaction.add(R.id.player_container, this.playerFragment!!, PlayerFragment.TAG)
                }
            }
        }
        transaction.commitNow()
    }

    private fun switchToHomeFragment() {
        if (this.currentFragment != Constants.Fragment.HOME) {
            val transaction = supportFragmentManager.beginTransaction()
            when {
                this.homeFragment != null -> transaction.show(this.homeFragment!!)
                else -> {
                    this.homeFragment = supportFragmentManager.findFragmentByTag(HomeFragment.TAG) as HomeFragment?
                    if (this.homeFragment != null) {
                        transaction.show(this.homeFragment!!)
                    } else {
                        this.homeFragment = HomeFragment()
                        this.homeFragment!!.onPlayItemsListener = this
                        transaction.add(R.id.content_frame, this.homeFragment!!, HomeFragment.TAG)
                    }
                }
            }
            this.searchFragment?.let { transaction.hide(this.searchFragment!!) }
            this.libraryFragment?.let { transaction.hide(this.libraryFragment!!) }
            transaction.commitNow()
            this.currentFragment = Constants.Fragment.HOME
            this.enableScroll()
        }
    }

    private fun switchToSearchFragment() {
        if (this.currentFragment != Constants.Fragment.SEARCH) {
            val transaction = supportFragmentManager.beginTransaction()
            when {
                this.searchFragment != null -> transaction.show(this.searchFragment!!)
                else -> {
                    this.searchFragment = supportFragmentManager.findFragmentByTag(SearchFragment.TAG) as SearchFragment?
                    if (this.searchFragment != null) {
                        transaction.show(this.searchFragment!!)
                    } else {
                        this.searchFragment = SearchFragment()
                        this.searchFragment!!.onPlayItemsListener = this
                        transaction.add(R.id.content_frame, this.searchFragment!!, SearchFragment.TAG)
                    }
                }
            }
            this.homeFragment?.let { transaction.hide(this.homeFragment!!) }
            this.libraryFragment?.let { transaction.hide(this.libraryFragment!!) }
            transaction.commitNow()
            this.currentFragment = Constants.Fragment.SEARCH
            this.disableScroll()
        }
    }

    private fun switchToLibraryFragment() {
        if (this.currentFragment != Constants.Fragment.LIBRARY) {
            val transaction = supportFragmentManager.beginTransaction()
            when {
                this.libraryFragment != null -> transaction.show(this.libraryFragment!!)
                else -> {
                    this.libraryFragment = supportFragmentManager.findFragmentByTag(LibraryFragment.TAG) as LibraryFragment?
                    if (this.libraryFragment != null) {
                        transaction.show(this.libraryFragment!!)
                    } else {
                        this.libraryFragment = LibraryFragment()
                        this.libraryFragment!!.onPlayItemsListener = this
                        transaction.add(R.id.content_frame, this.libraryFragment!!, LibraryFragment.TAG)
                    }
                }
            }
            this.homeFragment?.let { transaction.hide(this.homeFragment!!) }
            this.searchFragment?.let { transaction.hide(this.searchFragment!!) }
            transaction.commitNow()
            this.currentFragment = Constants.Fragment.LIBRARY
            this.enableScroll()
        }
    }

    private fun refreshVideoFavorited() {
        this.playerFragment?.let {
            val currentVideo = playerFragment!!.currentVideo
            currentVideo?.let { playerFragment!!.updateVideo(GsonPlaylistHelper.isFavorited(currentVideo)) }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(R.style.AppTheme)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        GsonPlaylistHelper.create(this)

        this.currentFragment = -1

        startService(Intent(this, KillNotificationService::class.java))

        setupReceivers()
        setSupportActionBar(toolbar)
        bottom_nav_bar.setOnNavigationItemSelectedListener(this)
        switchToHomeFragment()
        setupPlayerFragment()

        onNewIntent(intent)
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        if (intent != null) {
            val action = intent.action
            if (action == Intent.ACTION_SEND) {
                val text = intent.getStringExtra(Intent.EXTRA_TEXT)
                val pattern = Pattern.compile(YOUTUBE_REGEX)
                val matcher = pattern.matcher(text)
                if (matcher.find()) {
                    val videoId = matcher.group(2)
                    if (videoId.length == 11) {
                        var video = Video(videoId)
                        video = YouTubeService.get().requestDetails(video)
                        onPlay(video)
                    }
                }
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_settings -> {
                startActivity(Intent(this, SettingsActivity::class.java))
                true
            }
            R.id.action_rate -> {
                this.launchMarket()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun launchMarket() {
        var uri = Uri.parse("market://details?id=$packageName")
        val action = Intent.ACTION_VIEW
        val flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED or Intent.FLAG_ACTIVITY_CLEAR_TOP
        var marketIntent: Intent
        try {
            marketIntent = Intent(action, uri)
            marketIntent.setPackage("com.android.vending")
            marketIntent.flags = flags
            startActivity(marketIntent)
        } catch (e: ActivityNotFoundException) {
            uri = Uri.parse("https://play.google.com/store/apps/details?id=$packageName")
            marketIntent = Intent(action, uri)
            marketIntent.flags = flags
            startActivity(marketIntent)
        }

    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_toolbar, menu)
        return true
    }

    override fun onBackPressed() {
        if (!this.playerFragment!!.swipeDown()) {
            bottom_nav_bar.menu.findItem(R.id.nav_item_home).isChecked = true
            this.switchToHomeFragment()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        unregisterReceiver(this.playReceiver)
        unregisterReceiver(this.wifiReceiver)
    }

    override fun onPlayAll(videos: Collection<Video>) {
        if (playerFragment != null && videos.isNotEmpty() && playerFragment!!.setQueueTo(videos)) {
            playerFragment!!.forcePlayNext()
        }
    }

    override fun onPlay(video: Video) {
        if (playerFragment != null && playerFragment!!.addToQueue(video)) {
            playerFragment!!.tryPlayNext()
        }
    }

    override fun onShuffle(videos: Collection<Video>) {
        onPlayAll(videos.shuffled())
    }

    override fun onPlayNext(video: Video) {
        if (playerFragment != null && playerFragment!!.addToQueue(0, video)) {
            playerFragment!!.tryPlayNext()
        }
    }

    override fun onPlayNow(video: Video) {
        if (playerFragment != null && playerFragment!!.addToQueue(0, video)) {
            playerFragment!!.forcePlayNext()
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        this.playerFragment?.swipeDown()
        return when (item.itemId) {
            R.id.nav_item_home -> {
                switchToHomeFragment()
                true
            }
            R.id.nav_item_search -> {
                switchToSearchFragment()
                true
            }
            R.id.nav_item_library -> {
                switchToLibraryFragment()
                true
            }
            else -> false
        }
    }

    companion object {

        const val TAG = "MainActivity"

        const val YOUTUBE_REGEX = """^(?:(?:https?:\/\/)?(?:www\.)?)?(youtube(?:-nocookie)?\.com|youtu\.be)\/.*?(?:embed|e|v|watch\?.*?v=)?\/?([a-zA-Z0-9]+)"""

    }

}
